File: SSUMobile/Modules/Resources/SSUResourcesModule.m
//
//  SSUResourcesModule.m
//  SSUMobile
//
//  Created by Eric Amorde on 9/8/15.
//  Copyright (c) 2015 Sonoma State University Department of Computer Science. All rights reserved.
//

#import "SSUResourcesModule.h"
#import "SSUResourcesBuilder.h"
#import "SSUMoonlightCommunicator.h"
#import "SSULogging.h"
#import "SSUConfiguration.h"

@implementation SSUResourcesModule

+ (instancetype) sharedInstance {
    static SSUResourcesModule * instance = nil;
    static dispatch_once_t token;
    dispatch_once(&token, ^{
        instance = [[self alloc] init];
    });
    return instance;
}

#pragma mark - SSUModule

- (nonnull NSString *) title {
    return NSLocalizedString(@"Resources",
                             @"Campus resources, such as phone numbers and web links");
}

- (nonnull NSString *) identifier {
    return @"directory";
}

- (UIView *) viewForHomeScreen {
    return [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"resources_icon"]];
}

- (UIImage *) imageForHomeScreen {
    return [UIImage imageNamed:@"resources_icon"];
}

- (UIViewController *) initialViewController {
    UIStoryboard * storyboard = [UIStoryboard storyboardWithName:@"Resources"
                                                          bundle:[NSBundle bundleForClass:[self class]]];
    return [storyboard instantiateInitialViewController];
}

- (BOOL) showModuleInNavigationBar {
    return NO;
}

- (BOOL) shouldNavigateToModule {
    return YES;
}

- (void) setup {
    NSManagedObjectModel * model = [self modelWithName:@"Resources"];
    NSPersistentStoreCoordinator * coordinator = [self persistentStoreCoordinatorWithName:@"Resources" model:model];
    self.context = [self contextWithPersistentStoreCoordinator:coordinator];
    self.backgroundContext = [self backgroundContextFromContext:self.context];
}

- (void) clearCachedData {
    //TODO: implement clearCachedData
}

- (void) updateData:(void (^)())completion {
    SSULogDebug(@"Update Resources");
    [self updateSections:^{
        [self updateResources:^{
            if (completion) {
                completion();
            }
        }];
    }];
}


- (void) updateSections:(void (^)())completion {
    SSULogDebug(@"Update Resource Sections");
    NSDate * date = [[SSUConfiguration sharedInstance] dateForKey:SSUResourcesSectionsUpdatedDateKey];
    [SSUMoonlightCommunicator getJSONFromPath:@"ssumobile/resources/section" sinceDate:date completion:^(NSURLResponse * response, id json, NSError * error) {
        if (error != nil) {
            SSULogError(@"Error while attemping to update resource Sections: %@", error);
            if (completion) {
                completion();
            }
        }
        else {
            [[SSUConfiguration sharedInstance] setDate:[NSDate date] forKey:SSUResourcesSectionsUpdatedDateKey];
            [self.backgroundContext performBlock:^{
                [self buildSections:json];
                if (completion) {
                    completion();
                }
            }];
        }
        SSULogDebug(@"Finish %@",self.title);
    }];
}

- (void) updateResources:(void (^)())completion {
    SSULogDebug(@"Update Resource Sections");
    NSDate * date = [[SSUConfiguration sharedInstance] dateForKey:SSUResourcesResourcesUpdatedDateKey];
    [SSUMoonlightCommunicator getJSONFromPath:@"ssumobile/resources/resource" sinceDate:date completion:^(NSURLResponse * response, id json, NSError * error) {
        if (error != nil) {
            SSULogError(@"Error while attemping to update Resources: %@", error);
            if (completion) {
                completion();
            }
        }
        else {
            [[SSUConfiguration sharedInstance] setDate:[NSDate date] forKey:SSUResourcesResourcesUpdatedDateKey];
            [self.backgroundContext performBlock:^{
                [self buildResources:json];
                if (completion) {
                    completion();
                }
            }];
        }
        SSULogDebug(@"Finish %@",self.title);
    }];
}

#pragma mark - Private

- (void) buildResources:(id)json {
    SSUResourcesBuilder * builder = [[SSUResourcesBuilder alloc] init];
    builder.context = self.backgroundContext;
    [builder buildResources:json];
}


- (void) buildSections:(id)json {
    SSUResourcesBuilder * builder = [[SSUResourcesBuilder alloc] init];
    builder.context = self.backgroundContext;
    [builder buildSections:json];
}

@end

File: SSUMobile/Modules/Resources/SSUResourcesModule.swift
//
//  SSUResourcesModule.swift
//  SSUMobile
//
//  Created by Eric Amorde on 4/8/17.
//  Copyright © 2017 Sonoma State University Department of Computer Science. All rights reserved.
//

import Foundation

class SSUResourcesModule: SSUCoreDataModuleBase, SSUModuleUI {
        
    @objc(sharedInstance)
    static let instance = SSUResourcesModule()
    
    let articleFetchDateLimit: TimeInterval = 60*60*24*180; // 6 months
    
    // MARK: SSUModule
    
    var title: String {
        return NSLocalizedString("Resources",
                                 comment: "Campus resources, such as phone numbers and web links")
    }
    
    var identifier: String {
        return "resources"
    }
    
    func setup() {
        setupCoreData(modelName: "Resources", storeName: "Resources")
    }
    
    func updateData(_ completion: (() -> Void)? = nil) {
        SSULogging.logDebug("Update Resources");
        updateSections {
            self.updateResources {
                completion?()
            }
        }
    }
    
    func updateSections(completion: (() -> Void)? = nil) {
        SSULogging.logDebug("Update Resource Sections");
        let lastUpdate = SSUConfiguration.sharedInstance().date(forKey: SSUResourcesSectionsUpdatedDateKey)
        SSUMoonlightCommunicator.getJSONFromPath("ssumobile/resources/section", since: lastUpdate) { (response, json, error) in
            if let error = error {
                SSULogging.logError("Error while updating Resources: \(error)")
                completion?()
            } else {
                SSUConfiguration.sharedInstance().setDate(Date(), forKey: SSUResourcesSectionsUpdatedDateKey)
                self.buildSections(json: json)
                completion?()
            }
        }
        
        SSULogging.logDebug("Finish \(title)")
    }
    
    func updateResources(completion: (() -> Void)? = nil) {
        SSULogging.logDebug("Update Resource resources");
        let lastUpdate = SSUConfiguration.sharedInstance().date(forKey: SSUResourcesResourcesUpdatedDateKey)
        SSUMoonlightCommunicator.getJSONFromPath("ssumobile/resources/resource", since: lastUpdate) { (response, json, error) in
            if let error = error {
                SSULogging.logError("Error while updating Resources: \(error)")
                completion?()
            } else {
                SSUConfiguration.sharedInstance().setDate(Date(), forKey: SSUResourcesResourcesUpdatedDateKey)
                self.buildResources(json: json)
                completion?()
            }
        }
    }
    
    private func buildResources(json: Any) {
        guard let jsonData = json as? [Any] else {
            SSULogging.logError("Incorrect data type returned from resources")
            return
        }
        let builder = SSUResourcesBuilder()
        builder.context = backgroundContext
        backgroundContext.perform {
            builder.buildResources(jsonData)
        }
    }
    
    private func buildSections(json: Any) {
        guard let jsonData = json as? [Any] else {
            SSULogging.logError("Incorrect data type returned from resources sections")
            return
        }
        let builder = SSUResourcesBuilder()
        builder.context = backgroundContext
        backgroundContext.perform {
            builder.buildSections(jsonData)
        }
    }
    
    // MARK: SSUModuleUI
    
    func imageForHomeScreen() -> UIImage? {
        return UIImage(named: "resources_icon")
    }
    
    func viewForHomeScreen() -> UIView? {
        return nil
    }
    
    func initialViewController() -> UIViewController {
        let storyboard = UIStoryboard(name: "Resources", bundle: Bundle(for: type(of: self)))
        return storyboard.instantiateInitialViewController()!
    }
    
    func shouldNavigateToModule() -> Bool {
        return true
    }
    
    func showModuleInNavigationBar() -> Bool {
        return false
    }
}
